aa-sdk/core/src/middleware/defaults/7702gasEstimator.ts

```diff
@@ -0,0 +1,44 @@
+import { AccountNotFoundError } from "../../errors/account.js";
+import type { UserOperationStruct } from "../../types.js";
+import type { ClientMiddlewareFn } from "../types";
+import { defaultGasEstimator } from "./gasEstimator.js";
+
+/**
+ * A middleware function to estimate the gas usage of a user operation with an optional custom gas estimator.
+ * This function is only compatible with accounts using EntryPoint v0.7.0.
+ *
+ * @param {ClientMiddlewareFn} [gasEstimator] An optional custom gas estimator function
+ * @returns {Function} A function that takes user operation structure and parameters, estimates gas usage, and returns the estimated user operation
+ */
+export const default7702GasEstimator: (
+  gasEstimator?: ClientMiddlewareFn
+) => ClientMiddlewareFn =
+  (gasEstimator?: ClientMiddlewareFn) => async (struct, params) => {
+    const gasEstimator_ = gasEstimator ?? defaultGasEstimator(params.client);
+
+    const account = params.account ?? params.client.account;
+    if (!account) {
+      throw new AccountNotFoundError();
+    }
+
+    const entryPoint = account.getEntryPoint();
+    if (entryPoint.version !== "0.7.0") {
+      throw new Error(
+        "This middleware is only compatible with EntryPoint v0.7.0"
+      );
+    }
+
+    // todo: this is currently overloading the meaning of the getImplementationAddress method, replace with a dedicated method or clarify intention in docs
```

Comment:
discussed offline, but I think it's a good idea to allow passing an optional override to `toSmartContractAccount` for this

That methods intention is meant to return the address at which the logic for the SCA lives and I think that holds here:
- proxies point to a different address than the SCA address
- no proxy accounts should probably return the address of the SCA itself?
- 7702 returns the address containing the logic we're delegating to for the EOA

Reply:
Added the override to `toSmartConctractAccount`. Details on the specifics of the `getImplementationAddress` added in the other GH review comment.